Implement memoization in several selectors and hooks

Aljullu opened the pull request:
Fixes #3855.

This PR implements some techniques to avoid unnecessary re-renders and side effects when using some of our selectors and hooks.

### How to test the changes in this Pull Request:

#### For devs (don't include this in the release testing steps)

It's possible to verify that some effects are no longer re-run on each interaction. For example:

1. Add `console.log( 'effect run' );` before [`context/providers/cart-checkout/shipping/index.js` L121](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/b8451c90d68be626ba9a41c71a5629a600fe5c72/assets/js/base/context/providers/cart-checkout/shipping/index.js#L121):

https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/b8451c90d68be626ba9a41c71a5629a600fe5c72/assets/js/base/context/providers/cart-checkout/shipping/index.js#L121

2. Go to the Cart block in the frontend.
3. Increase or decrease the quantity of a product.
4. Verify `effect run` is not logged every time.
5. If you repeat the test in trunk, you will notice `effect run` is logged every time the quantity changes.

#### For users (include this in the release testing steps)

1. Smoke test the Cart and Checkout blocks: try removing products, changing their quantity, modifying shipping address, adding coupons, etc.
2. Verify there are no regressions and everything is working fine.

### Changelog

> Cart and Checkout block performance improvements

Aljullu committed:
Memoize cartErrors from useStoreCart

Aljullu committed:
Implement memoization to other selectors

Aljullu committed:
Use EMPTY_ARRAY in more cases

Aljullu committed:
Create getArrayOrEmptyConstant function

Aljullu committed:
Remove DEFAULT_EMPTY_ARRAY constant and use EMPTY_ARRAY instead

Aljullu committed:
Wrap removeItem into useCallback

Aljullu committed:
Fix tests

ralucaStan commented on assets/js/base/context/hooks/cart/use-store-cart.ts:
```diff
@@ -173,39 +198,49 @@ export const useStoreCart = (
 
 			return {
 				cartCoupons,
-				cartItems: cartData.items || [],
```
Is cartData.items ever `undefined` or `null`? Do we ever set it or is this just a safety measurement? From what I saw in the state we always start with `[]`, so `|| []` is never reached. Which means that this fix here is also never reached.

ralucaStan replied:
Same observation about the changes below regarding cartData.errors, cartData.extensions, cartData.shippingRates and cartData.paymentRequirements.

Aljullu replied:
That's a good point... I investigated this and I think those default values were only needed for tests to pass, which means that the issue was in the tests configuration and not in the code itself.

I made some changes in `assets/js/base/context/hooks/cart/test/use-store-cart.js` and `assets/js/previews/cart.ts` and now tests pass without requiring those defaults values to exist.

ralucaStan replied:
Congrats on investigating this.

ralucaStan commented on assets/js/data/cart/selectors.ts:
```diff
@@ -46,7 +47,7 @@ export const getCartMeta = ( state: CartState ): CartMeta => {
 export const getCartErrors = ( state: CartState ): Array< ResponseError > => {
-	return state.errors || [];
+	return state.errors || EMPTY_ARRAY;
 };
```
I think using a global EMPTY_ARRAY is really error prone and could affect all the data in the store, as it could easily be mutated. If in the end it turns out that this is the way to go, I would recomment using something like EMPTY_CART_ERRORS instead of EMPTY_ARRAY:

```
const EMPTY_CART_ERRORS = [];

export const getCartErrors = ( state: CartState ): Array< ResponseError > => {
	return state.errors || [];
	return state.errors || EMPTY_CART_ERRORS;
};
```

- a localized variable, instead of a global one. We don't export it and there is no danger of sharing it among data.
- a specific variable for a specific data branch

Aljullu replied:
I changed all `EMPTY_ARRAY` and `EMPTY_OBJECT` instances for localized variables, that means there are many more variables now, but I think it's ok. :ok_hand:

ralucaStan reviewed (commented):
From what I understood from @Aljullu, _when we return an empty array from our selectors, any component that is subscribed to them and uses them as a useEffect() or useCallback() dependency, will be running that effect or callback on every render (because [] === [] //false)._

I am curious why we need to return empty [] or {}. From what I noticed our store has all the cart data already initialized to empty values, like '', [] or {}.
For example, when we select the data `cartData.errors || []`, cartData.errors is not `undefined` nor `null`, so that empty array is never reached. 

So I need a bit of context of why we actually need to return in our selectors || [] or || {}, because maybe the problem lies somewhere else, in the place where we actually set a cart data to null/undefined instead of resetting it to the default cart value for that data,  like '', [] or {}.

When it comes to using this EMPTY_OBJECT solution I think we should be really careful, because this introduces a high degree of risk. It's a global variable, shared across multiple data with a high risk of mutation. I see a lot of places where EMPTY_OBJECT is used and this could also be something that we might end up doing without really understanding the need.  I presented in a comment a solution with something localized, more specific see EMPTY_CART_ERRORS. 



Lastly, I want us to comment as a team about what we are trying to solve here. Do we have a really big performance issue that is worth bringing this amount of unpredictability into the project?

Aljullu committed:
Do not default to empty array/object for values that should exist and use different constant for empty values

Aljullu commented:
@ralucaStan I updated this PR addressing your feedback, let me know how it looks now.

> Lastly, I want us to comment as a team about what we are trying to solve here. Do we have a really big performance issue that is worth bringing this amount of unpredictability into the project?

I don't think there is a performance issue we want to fix specifically, I just think this is a good practice to implement even if it doesn't have a real performance impact. Another thing to consider is that some of those values might be used by extensions, so even if in WC Blocks there isn't a performance issue, there might be in 3rd-party extensions relying on those values.

ralucaStan commented on assets/js/data/constants.ts:
```diff
@@ -0,0 +1,16 @@
+export const EMPTY_SHIPPING_RATES: [  ] = [];
+export const EMPTY_PAYMENT_REQUIREMENTS: [  ] = [];
+export const EMPTY_EXTENSIONS: Record< string, unknown > = {};
+export const EMPTY_TAX_LINES: [  ] = [];
```
👏 this actually triggers a warning when trying to modify any of this values

ralucaStan commented on assets/js/data/default-states.ts:
```diff
@@ -16,12 +27,15 @@ export interface CartState {
 	errors: Array< ResponseError >;
 }
 
+export const EMPTY_PENDING_QUANTITY = [];
```
Let's type these as well.
```
export const EMPTY_PENDING_QUANTITY : [ ] = [];
export const EMPTY_PENDING_DELETE : [ ] = [];
```

Aljullu replied:
Good catch! This is done in 8282985fbfea5dd37babd55694cc36121ce8e527.

ralucaStan commented on assets/js/base/context/hooks/cart/test/use-store-cart.js:
```diff
@@ -17,6 +17,7 @@ jest.mock( '../../../providers/editor-context', () => ( {
 } ) );
 
 jest.mock( '@woocommerce/block-data', () => ( {
+	...jest.requireActual( '@woocommerce/block-data' ),
```
What is the purpose of this? Is this something we should do for other mocks in our tests in another PR, I see not all of them use them.

Aljullu replied:
If this line wasn't there, any exported constant that wasn't mocked, would be undefined. In this case, all new `EMPTY_XYZ` arrays would be undefined because they are not explicitly added to the mock.

I'm not sure if there is any drawback of doing this change to all other mocks, but it makes tests to be closer to the way production code runs, so in general I think it's better to have it, even if not needed.

ralucaStan reviewed (approved):
I've left a few comments and requested a small change. After that is addressed this could be merged from my side

Aljullu committed:
Type two empty array constants from default-states.ts
